File: golang/dates_test.go
package main

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
)

var (
	now                = time.Now()
	sevenDaysAgo       = now.AddDate(0, 0, -7)
	sevenDaysFuture    = now.AddDate(0, 0, 7)
	fourteenDaysFuture = now.AddDate(0, 0, 14)
)

func TestRange1ExtendsIntoRange2(t *testing.T) {
	var dateRange1 = DateRange{sevenDaysAgo, sevenDaysFuture}
	var dateRange2 = DateRange{now, fourteenDaysFuture}
	var expected = DateRange{now, sevenDaysFuture}
	result, _ := findOverlap(dateRange1, dateRange2)

	assert.Equal(t, expected, result, "Unexpected DateRange")
}

func TestRange2ExtendsIntoRange1(t *testing.T) {
	var dateRange2 = DateRange{sevenDaysAgo, sevenDaysFuture}
	var dateRange1 = DateRange{now, fourteenDaysFuture}
	var expected = DateRange{now, sevenDaysFuture}
	result, _ := findOverlap(dateRange1, dateRange2)

	assert.Equal(t, expected, result, "Unexpected DateRange")
}

func TestRange1EndsAtInstantRange2Begins(t *testing.T) {
	var dateRange1 = DateRange{sevenDaysAgo, now}
	var dateRange2 = DateRange{now, fourteenDaysFuture}
	_, err := findOverlap(dateRange1, dateRange2)

	assert.NotNil(t, err, "Expected an error")
}

func TestRange2EndsAtInstantRange1Begins(t *testing.T) {
	var dateRange2 = DateRange{sevenDaysAgo, now}
	var dateRange1 = DateRange{now, fourteenDaysFuture}
	_, err := findOverlap(dateRange1, dateRange2)

	assert.NotNil(t, err, "Expected an error")
}

func TestRange1ContainsRange2(t *testing.T) {
	var dateRange1 = DateRange{sevenDaysAgo, fourteenDaysFuture}
	var dateRange2 = DateRange{now, sevenDaysFuture}
	var expected = DateRange{now, sevenDaysFuture}
	result, _ := findOverlap(dateRange1, dateRange2)

	assert.Equal(t, expected, result, "Unexpected DateRange")
}

func TestRange2ContainsRange1(t *testing.T) {
	var dateRange2 = DateRange{sevenDaysAgo, fourteenDaysFuture}
	var dateRange1 = DateRange{now, sevenDaysFuture}
	var expected = DateRange{now, sevenDaysFuture}
	result, _ := findOverlap(dateRange1, dateRange2)

	assert.Equal(t, expected, result, "Unexpected DateRange")
}

func TestRange1DoesNotOverlapRange2(t *testing.T) {
	var dateRange1 = DateRange{sevenDaysAgo, now}
	var dateRange2 = DateRange{sevenDaysFuture, fourteenDaysFuture}
	_, err := findOverlap(dateRange1, dateRange2)

	assert.NotNil(t, err, "Expected an error")
}

func TestRange2DoesNotOverlapRange1(t *testing.T) {
	var dateRange2 = DateRange{sevenDaysAgo, now}
	var dateRange1 = DateRange{sevenDaysFuture, fourteenDaysFuture}
	_, err := findOverlap(dateRange1, dateRange2)

	assert.NotNil(t, err, "Expected an error")
}

func TestRangesIdentical(t *testing.T) {
	var dateRange2 = DateRange{sevenDaysAgo, fourteenDaysFuture}
	var dateRange1 = DateRange{sevenDaysAgo, fourteenDaysFuture}
	var expected = DateRange{sevenDaysAgo, fourteenDaysFuture}
	result, _ := findOverlap(dateRange1, dateRange2)

	assert.Equal(t, expected, result, "Unexpected DateRange")
}

File: ts/dates.test.ts
import { expect } from 'chai';
import { findOverlap } from './dates';

describe('findOverlap', () => {
    const now = new Date();
    const sevenDaysAgo = addDays(now, -7);
    const sevenDaysFuture = addDays(now, 7);
    const fourteenDaysFuture = addDays(now, 14);

    it('should return overlap when range one extends into range two', () => {
        const dateRange1 = { begin: sevenDaysAgo, end: sevenDaysFuture };
        const dateRange2 = { begin: now, end: fourteenDaysFuture };
        const expected = { begin: now, end: sevenDaysFuture };
        const result = findOverlap(dateRange1, dateRange2);

        expect(expected.begin.valueOf()).to.equal(result.begin.valueOf());
        expect(expected.end.valueOf()).to.equal(result.end.valueOf());
    });

    it('should return overlap when range two extends into range one', () => {
        const dateRange2 = { begin: sevenDaysAgo, end: sevenDaysFuture };
        const dateRange1 = { begin: now, end: fourteenDaysFuture };
        const expected = { begin: now, end: sevenDaysFuture };
        const result = findOverlap(dateRange1, dateRange2);

        expect(expected.begin.valueOf()).to.equal(result.begin.valueOf());
        expect(expected.end.valueOf()).to.equal(result.end.valueOf());
    });

    it('should throw exception when range one ends at instant range two begins', () => {
        const dateRange1 = { begin: sevenDaysAgo, end: now };
        const dateRange2 = { begin: now, end: fourteenDaysFuture };
        const func = () => findOverlap(dateRange1, dateRange2);

        expect(func).to.throw();
    });

    it('should throw exception when range two ends at instant range one begins', () => {
        const dateRange2 = { begin: sevenDaysAgo, end: now };
        const dateRange1 = { begin: now, end: fourteenDaysFuture };
        const func = () => findOverlap(dateRange1, dateRange2);

        expect(func).to.throw();
    });

    it('should return overlap when range one completely contains range two', () => {
        const dateRange1 = { begin: sevenDaysAgo, end: fourteenDaysFuture };
        const dateRange2 = { begin: now, end: sevenDaysFuture };
        const expected = { begin: now, end: sevenDaysFuture };
        const result = findOverlap(dateRange1, dateRange2);

        expect(expected.begin.valueOf()).to.equal(result.begin.valueOf());
        expect(expected.end.valueOf()).to.equal(result.end.valueOf());
    });

    it('should return overlap when range two completely contains range one', () => {
        const dateRange2 = { begin: sevenDaysAgo, end: fourteenDaysFuture };
        const dateRange1 = { begin: now, end: sevenDaysFuture };
        const expected = { begin: now, end: sevenDaysFuture };
        const result = findOverlap(dateRange1, dateRange2);

        expect(expected.begin.valueOf()).to.equal(result.begin.valueOf());
        expect(expected.end.valueOf()).to.equal(result.end.valueOf());
    });

    it('should throw exception when range one does not overlap with range two', () => {
        const dateRange1 = { begin: sevenDaysAgo, end: now };
        const dateRange2 = { begin: sevenDaysFuture, end: fourteenDaysFuture };
        const func = () => findOverlap(dateRange1, dateRange2);

        expect(func).to.throw();
    });

    it('should throw exception when range two does not overlap range one', () => {
        const dateRange2 = { begin: sevenDaysAgo, end: now };
        const dateRange1 = { begin: sevenDaysFuture, end: fourteenDaysFuture };
        const func = () => findOverlap(dateRange1, dateRange2);

        expect(func).to.throw();
    });

    it('should return overlap when range one equals range two', () => {
        const dateRange1 = { begin: sevenDaysAgo, end: fourteenDaysFuture };
        const dateRange2 = { begin: sevenDaysAgo, end: fourteenDaysFuture };
        const expected = { begin: sevenDaysAgo, end: fourteenDaysFuture };
        const result = findOverlap(dateRange1, dateRange2);

        expect(expected.begin.valueOf()).to.equal(result.begin.valueOf());
        expect(expected.end.valueOf()).to.equal(result.end.valueOf());
    });
});

function addDays(date: Date, days: number) {
    var result = new Date(date.getTime());
    result.setDate(result.getDate() + days);
    return result;
}